data-processing/entities/terms/extractor.py

```diff
@@ -10,110 +9,169 @@
-                # Assuming DEFAULT_CONTEXT_SIZE is 20 characters
-                context_tex=tex[start - \
-                                DEFAULT_CONTEXT_SIZE: end + DEFAULT_CONTEXT_SIZE]
-            )
-            index += 1
+	"""
+	Extract plaintext terms from TeX, with the locations of the start and end characters they correspond to in
+	the input TeX strings.
+	"""
+
+	# Extracting all the terms, definition_text and definition_html from glossary
+	terms_and_def_text_html = []
+
+	for obj in glossary:
+		new_triple = []
+
+		#Making terms lower case in one step
+		new_triple.append(obj["terms"].lower())
+		new_triple.append(obj["definition_text"])
+		new_triple.append(obj["definition_html"])
+		terms_and_def_text_html.append(new_triple)
+
+	matched_terms = []
+	matched_definition_texts = []
+	matched_definition_htmls = []
+	matched_terms_start = []
+	matched_terms_end = []
+	matched_terms_context_str = []
+
+	def parse(self, tex_path: str, tex: str) -> Iterator[Term]:
+		# Extract plaintext segments from TeX
+		plaintext_extractor = PlaintextExtractor()
+		plaintext_segments = plaintext_extractor.parse(tex_path, tex)
+
+		global terms_and_def_text_html
+		global matched_terms
+		global matched_definition_texts
+		global matched_definition_htmls
+		global matched_terms_start
+		global matched_terms_end
+		global matched_terms_context_str
+
+		for text_segment in plaintext_segments:
+			text_segment_string = str(text_segment.text.lower())
+
+			# Making two lists from the text_segment_string; one is to capture words without dashes
+			# The other is to capture words such as "Gradient-Activated" (but this word happens to not be
+			# in the glossary
+			text_segment_string_split_on_dashes = re.split(
+				"[-\s+]", text_segment_string)
+			text_segment_string_not_split_on_dashes = re.split(
+				"[\s+]", text_segment_string)
+
+			# Adding all terms that were both in the glossary and in the .tex file into a
+			# set so we don't get duplicates
+			matches_tuples = []
+			matches = set()
+			for word in text_segment_string_split_on_dashes:
+				for object in terms_and_def_text_html:
+
+					# Object[0] is the term
+					if object[0] == word:
+
+						# Adding term name, text_segment from which term was found, def text and def html
+						matches_tuples.append([object[0], text_segment_string, object[1], object[2]])
+
+			for word2 in text_segment_string_not_split_on_dashes:
+				for object in terms_and_def_text_html:
+					if object[0] == word2:
+						matches_tuples.append([object[0], text_segment_string, object[1], object[2]])
+			for object in terms_and_def_text_html:
+				term_list = re.split("\s+", object[0])
+				if len(term_list) > 1:
+					if object[0] in text_segment_string:
+						matches_tuples.append([object[0], text_segment_string, object[1], object[2]])
+
+			for i in matches_tuples:
+				list_to_add = i[:-1]
+				if isinstance(i[-1], list):
+					last_elem = tuple(i[-1])
```

Comment:
Why does an element in the match tuples need to be cast from a list to a tuple here?

Reply:
I wanted to add it to the matches set. Python won't allow lists to be added to set objects, but it allows tuples to be added.